### PR TITLE
Add TemplateCodeInfo for code info in job template only

### PR DESCRIPTION
**Pull request by unkcpz**
This is another piece required for containerized code implementation. In the current design, the code_info of calc_job is read from the code setup and plugin then pass to the job template to create the bash script. 
However, job template needs more flexibility to control the different part of script runline where currently all the part (1. exec_name from code uuid, 2. code_info.cmdline_params, 3. mpi parameters from computer setting) are stacked together to the job template's code_info. 

In this PR, the class `TemplateCodeInfo` is created to handle the elements needed to generate a job script. Where the `code_uuid` and `withmpi` fields are not used in job script generation.

### CODE REVIEW

File: aiida/schedulers/datastructures.py [hunk 1]
```diff
@@ -352,6 +352,20 @@ class JobTemplate(DefaultFieldsAttributeDict):  # pylint: disable=too-many-insta
...
+    """
+    _default_fields = (
+        'cmdline_params',  # as a list of strings
+        'stdin_name',
+        'stdout_name',
+        'stderr_name',
+        'join_files',
+    )
+
+
```
Comment (chrisjsewell) [inline aiida/schedulers/datastructures.py:368]: Heya, lets move away from using these bespoke data class solutions, and use python's built in solution: https://docs.python.org/3/library/dataclasses.html

```suggestion
from __future__ import annotations
from dataclasses import dataclass, field

@dataclass
class TemplateCodeInfo):
    """
    Extra layer of code_info which parse code info from code and plugin setting and then
    pass to job template
    """
    cmdline_params: list[str] = field(default_factory=list)
    stdin_name: None | str = None
    stdout_name: None | str = None
    stderr_name: None | str = None
    join_files: list[str] = field(default_factory=list)


```

Comment (unkcpz) [reply aiida/schedulers/datastructures.py:368]: Cool, I like this style. Is this a newly introduced feature?

Comment (chrisjsewell) [reply aiida/schedulers/datastructures.py:368]: Yep, from python 3.7+, and we are python>=3.8

Comment (unkcpz) [reply aiida/schedulers/datastructures.py]: done. Thanks @chrisjsewell

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @unkcpz . Have a few suggestions. The tests are failing because the `JobTemplate` that is generated by the engine is written to the repository of the `CalcJobNode` by dumping it to JSON. The old `CodeInfo` was a simple `dict` and so serializable, but the new `dataclasses` are not JSON serializable.
```python
In [1]: import json

In [2]: from dataclasses import dataclass

In [4]: @dataclass
   ...: class Test:
   ...:     a: int = 1
   ...: 

In [5]: a = Test()

In [6]: a.a
Out[6]: 1

In [7]: json.dumps(a)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-2f50cf32d976> in <module>
----> 1 json.dumps(a)

/usr/lib/python3.9/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    229         cls is None and indent is None and separators is None and
    230         default is None and not sort_keys and not kw):
--> 231         return _default_encoder.encode(obj)
    232     if cls is None:
    233         cls = JSONEncoder

/usr/lib/python3.9/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

/usr/lib/python3.9/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

/usr/lib/python3.9/json/encoder.py in default(self, o)
    177 
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')
    181 

TypeError: Object of type Test is not JSON serializable
```

However, you can turn it into a dictionary first. So in line 763, of `CalcJob.presubmit` turn
```python
subfolder.create_file_from_filelike(io.StringIO(json.dumps(job_tmpl)), 'job_tmpl.json', 'w', encoding='utf8')
```
into
```python
from dataclasses import asdict
subfolder.create_file_from_filelike(io.StringIO(json.dumps(asdict(job_tmpl))), 'job_tmpl.json', 'w', encoding='utf8')
```

**Comment by unkcpz** [comment]
Thanks @sphuber. I check the failed test caused by `dataclass`. I think we have to work around it with `AttributeDict(asdict(tmpl_code_info))`. It is `tmpl_code_info` who is the `dataclass` and can be escaped to `dict` not `job_tmpl`. Converting to `dict` is not enough since the initial benifite from using `dataclass` is we have a built-in way to use dot to access attributes. Therefore need to convert to `AttributeDict` in order to do so. It seems not very concise but I don't know a more elegant solution.

**Comment by chrisjsewell** [comment]
> Therefore need to convert to `AttributeDict` in order to do so.

Err, I feel the better solution is just to handle data classes at JSON dump time:

```python
import dataclasses
def encoder(obj):
     if dataclasses.is_dataclass(obj):
        return  dataclasses.as_dict(obj) 
    raise TypeError(f" {obj!r} is not JSON serializable")
     
subfolder.create_file_from_filelike(io.StringIO(json.dumps(job_tmpl, default=encoder)), 'job_tmpl.json', 'w', encoding='utf8')
```

But, I don't want to overcomplicate the issue, so I'll leave it up to you guys, and maybe open this as a more general issue:

I'm very much not a fan of these `AttributeDict`, having personally made the "mistake" of implementing them in another one of my packages and then later removing them: https://github.com/executablebooks/markdown-it-py/issues/66.
One should use either dicts or dataclasses, but not mix the concepts.

@sphuber I think one "argument" for using them was dynamic auto-completion but, out of interest, I found that since  https://github.com/ipython/ipython/pull/5304, you can anyhow auto-complete dict keys:

image

image

**Comment by sphuber** [comment]
> For the `extendeddicts` classes, I think `AttributeDict` can quickly apply to the nested dict and make it dot visitable which is used in the `NodeLinksManager`. Is this also supported by `dataclasses`?

No, `dataclasses` do not recursively add attribute dereferencing for nested dicts that it contains. This is not what `dataclass`s are designed for though. I also don't think that it is recommended that its attributes are defined dynamically, as is the case for the `NodeLinksManager`, so we cannot or at least shouldn't use dataclasses for that.

The dataclasses should simply be used whenever we have a simple class that essentially just wraps some data through attributes. This is the case for `CodeInfo`, `CalcInfo` etc.

**Comment by sphuber** [comment]
@giovannipizzi @chrisjsewell even though the chances are small, this *could* potentially be breaking if there are scheduler plugins that rely on the type of the `codes_info` argument passed to the `_get_run_line` method. Do we accept these chances and put this in 2.0 still?

**Comment by ltalirz** [comment]
> @giovannipizzi @chrisjsewell even though the chances are small, this _could_ potentially be breaking if there are scheduler plugins that rely on the type of the `codes_info` argument passed to the `_get_run_line` method. Do we accept these chances and put this in 2.0 still?

The AiiDA registry currently contains four packages that add a scheduler plugin:

https://github.com/zhubonan/aiida-fireworks-scheduler
https://github.com/pzarabadip/aiida-metavo-scheduler
https://github.com/aiidateam/aiida-firecrest
https://github.com/dev-zero/aiida-statefile-schedulers

@unkcpz Can you check whether any of those would be impacted?
If not, I think we can assume it is safe to proceed here (of course there can be private plugins that can be affected; we will just document the change in API).

**Comment by unkcpz** [comment]
@ltalirz thanks!

> @unkcpz Can you check whether any of those would be impacted?
Sure, I will go through them this week, if nothing particular is messed up by this PR, we can give it merged.

**Comment by unkcpz** [comment]
> The AiiDA registry currently contains four packages that add a scheduler plugin:
https://github.com/zhubonan/aiida-fireworks-scheduler
https://github.com/pzarabadip/aiida-metavo-scheduler
https://github.com/aiidateam/aiida-firecrest
https://github.com/dev-zero/aiida-statefile-schedulers

Just check all scheduler implementations. There are no codes that override the part of generating exec command of script. So it should be pretty safe to go forward. 😄

**Comment by ltalirz** [comment]
Great, any objections @sphuber ?

**Comment by sphuber** [comment]
No, if you rebase the PR and verify that it won't break the schedulers on the plugin registry, then I will approve and merge this.

**Comment by unkcpz** [comment]
Hi @sphuber, I rebase PR and check the `aiida-hyperqueue` is not impacted.

**Review by ltalirz — APPROVED** [review]
Thanks @unkcpz , my understanding is this can move forward !

**Review by sphuber — APPROVED** [review]

**Pull request merged**